File: nodejs_vs_golang/experiment.js
const { performance } = require('perf_hooks');
const os = require('os');

const numTasks = 100; // Jumlah tugas
const taskDuration = 100; // Durasi setiap tugas dalam milidetik

const runExperiment = async () => {
    const startTime = performance.now();

    // Simulasi tugas asinkron
    const tasks = Array.from({ length: numTasks }, (_, index) => new Promise((resolve) => {
        setTimeout(() => {
            resolve(`Task ${index + 1} completed`);
        }, taskDuration);
    }));

    // Tunggu semua tugas selesai
    await Promise.all(tasks);

    const endTime = performance.now();
    const elapsedTime = endTime - startTime; // Latency dalam milidetik

    // Hitung throughput
    const throughput = numTasks / (elapsedTime / 1000); // Tugas per detik

    console.log(`Total Time: ${elapsedTime.toFixed(2)} ms`);
    console.log(`Throughput: ${throughput.toFixed(2)} tasks/second`);
    console.log(`CPU Cores: ${os.cpus().length}`);
};

runExperiment();

File: nodejs_vs_golang/experiment.go
package main

import (
	"fmt"
	"runtime"
	"sync"
	"time"
)

const (
	numTasks     = 100                    // Jumlah tugas
	taskDuration = 100 * time.Millisecond // Durasi setiap tugas
)

func heavyOperation(id int, wg *sync.WaitGroup) {
	defer wg.Done()
	time.Sleep(taskDuration)
}

func main() {
	startTime := time.Now()

	var wg sync.WaitGroup
	wg.Add(numTasks)

	// Jalankan goroutine untuk setiap tugas
	for i := 1; i <= numTasks; i++ {
		go heavyOperation(i, &wg)
	}

	// Tunggu hingga semua goroutine selesai
	wg.Wait()

	endTime := time.Now()
	elapsedTime := endTime.Sub(startTime) // Latency

	// Hitung throughput
	throughput := float64(numTasks) / elapsedTime.Seconds()

	fmt.Printf("Total Time: %v\n", elapsedTime)
	fmt.Printf("Throughput: %.2f tasks/second\n", throughput)
	fmt.Printf("CPU Cores: %d\n", runtime.NumCPU())
}
